x/exchange/keeper/fulfillment.go

```diff
@@ -0,0 +1,375 @@
+func (k Keeper) FillAsks(ctx sdk.Context, msg *exchange.MsgFillAsksRequest) error {
+	if err := msg.ValidateBasic(); err != nil {
+		return err
+	}
+
+	marketID := msg.MarketId
+	store := k.getStore(ctx)
+
+	if err := validateAcceptingOrdersAndCanUserSettle(store, marketID); err != nil {
+		return err
+	}
+	buyer, serr := sdk.AccAddressFromBech32(msg.Buyer)
+	if serr != nil {
+		return fmt.Errorf("invalid buyer %q: %w", msg.Buyer, serr)
+	}
+	if err := k.validateUserCanCreateBid(ctx, marketID, buyer); err != nil {
+		return err
+	}
+	if err := validateCreateBidFees(store, marketID, msg.BidOrderCreationFee, msg.TotalPrice, msg.BuyerSettlementFees); err != nil {
+		return err
+	}
+
+	orders, oerrs := k.getAskOrders(store, marketID, msg.AskOrderIds, msg.Buyer)
+	if oerrs != nil {
+		return oerrs
+	}
+
+	var errs []error
+	assetInputs := make([]banktypes.Input, 0, len(msg.AskOrderIds))
+	priceOutputs := make([]banktypes.Output, 0, len(msg.AskOrderIds))
+	addrIndex := make(map[string]int, len(msg.AskOrderIds))
+	feeInputs := make([]banktypes.Input, 0, len(msg.AskOrderIds)+1)
+	feeAddrIndex := make(map[string]int, len(msg.AskOrderIds))
+	filledOrders := make([]*exchange.FilledOrder, 0, len(msg.AskOrderIds))
+	for _, order := range orders {
+		askOrder := order.GetAskOrder()
+		seller := askOrder.Seller
+		assets := askOrder.Assets
+		price := askOrder.Price
+		sellerSettlementFlatFee := askOrder.SellerSettlementFlatFee
+
+		sellerRatioFee, rerr := calculateSellerSettlementRatioFee(store, marketID, price)
+		if rerr != nil {
+			errs = append(errs, fmt.Errorf("error calculating seller settlement ratio fee for order %d: %w",
+				order.OrderId, rerr))
+			continue
+		}
+		if err := k.releaseHoldOnOrder(ctx, order); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		var totalSellerFee sdk.Coins
+		if sellerSettlementFlatFee != nil && !sellerSettlementFlatFee.IsZero() {
+			totalSellerFee = totalSellerFee.Add(*sellerSettlementFlatFee)
+		}
+		if sellerRatioFee != nil && !sellerRatioFee.IsZero() {
+			totalSellerFee = totalSellerFee.Add(*sellerRatioFee)
+		}
+
+		i, seen := addrIndex[seller]
+		if !seen {
+			i = len(assetInputs)
+			addrIndex[seller] = i
+			assetInputs = append(assetInputs, banktypes.Input{Address: seller})
+			priceOutputs = append(priceOutputs, banktypes.Output{Address: seller})
+		}
+		assetInputs[i].Coins = assetInputs[i].Coins.Add(assets)
+		priceOutputs[i].Coins = priceOutputs[i].Coins.Add(price)
+
+		if !totalSellerFee.IsZero() {
+			j, known := feeAddrIndex[seller]
+			if !known {
+				j = len(feeInputs)
+				feeAddrIndex[seller] = j
+				feeInputs = append(feeInputs, banktypes.Input{Address: seller})
+			}
+			feeInputs[j].Coins = feeInputs[j].Coins.Add(totalSellerFee...)
+		}
+
+		filledOrders = append(filledOrders, exchange.NewFilledOrder(order, price, totalSellerFee))
+	}
+
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	totalAssets, totalPrice := sumAssetsAndPrice(orders)
```

Comment:
Same as FillBids, does it make sense to move this up before filledOrders since its validation.

Reply:
Yup. I think I put it down there because for the validation, only one of them is needed, and the other isn't needed until later, but that's a good piece of validation to put up front.